packages/viz/src/utils/print/data.js

```diff
@@ -178,101 +44,261 @@ function processBgRange(selectedDataByDate) {
   });
+}
 
-  const allBasals = _.reduce(
-    selectedDataByDate, (all, date) => (all.concat(_.get(date, "data.basal", []))), []
+function processBasalRange(dataByDate) {
+  const basals = _.reduce(
+    dataByDate,
+    (all, date) => (
+      all.concat(_.get(date, "data.basal", []))
+    ),
+    []
   );
   const rawBasalRange = extent(
-    allBasals,
+    basals,
     (d) => (_.max([_.get(d, "suppressed.rate", 0), d.rate]))
   );
   // multiply the max rate by 1.1 to add a little buffer so the highest basals
   // don't sit at the very top of the basal rendering area and bump into boluses
-  selected.basalRange = [0, rawBasalRange[1] * 1.1];
-
-  _.forEach(selected.dataByDate, (dateData) => {
-    const { bounds, data: { basal: basals } } = dateData;
-    for (let i = 0; i < basals.length; ++i) {
-      const basal = basals[i];
-      // trim the first and last basals to fit within the date's bounds
-      if (basal.utc < bounds[0]) {
-        basal.duration = basal.duration - (bounds[0] - basal.utc);
-        basal.utc = bounds[0];
-      }
-      if (i === basals.length - 1 && dateData.date !== selected.dateRange[1]) {
-        basal.duration = bounds[1] - basal.utc;
-      }
-      let nextBasal;
-      basal.subType = basal.deliveryType;
-      delete basal.deliveryType;
-      if (i !== basals.length - 1) {
-        nextBasal = basals[i + 1];
-        if ((basal.utc + basal.duration) !== nextBasal.utc) {
-          basal.discontinuousEnd = true;
-          nextBasal.discontinuousStart = true;
+  return [0, rawBasalRange[1] * 1.1];
+}
+
+/**
+ *
+ * @param {{duration:number;epoch:number;epochEnd:number;subType:string;discontinuousEnd:boolean;discontinuousStart:boolean;}[]} basals
+ * @param {[number,number]} bounds
+ */
+function updateBasalDiscontinuous(basals, bounds) {
+  if (basals.length < 1) {
+    return;
+  }
+  let prevBasal = null;
+  for (let i=0; i<basals.length; i++) {
+    const basal = basals[i];
+    // trim the first and last basals to fit within the date's bounds
+    if (basal.epoch < bounds[0]) {
+      basal.duration = basal.duration - (bounds[0] - basal.epoch);
+      basal.epoch = bounds[0];
+      basal.utc = basal.epoch;
+      basal.normalTime = new Date(basal.epoch).toISOString();
+    }
+    if (basal.epochEnd > bounds[1]) {
+      basal.duration = basal.duration - (basal.epochEnd - bounds[1]);
+      basal.epochEnd = basal.epoch + basal.duration;
+      basal.normalEnd = new Date(basal.epochEnd).toISOString();
+    }
+
+    basal.discontinuousEnd = false;
+    basal.discontinuousStart = false;
+
+    if (prevBasal && (prevBasal.utc + prevBasal.duration) !== basal.utc) {
+      prevBasal.discontinuousEnd = true;
+      basal.discontinuousStart = true;
+    }
+
+    prevBasal = basal;
+  }
+}
+
+/**
+ *
+ * @param {import("tideline").TidelineData} tidelineData
+ * @param {moment.Moment} startDate
+ * @param {moment.Moment} endDate
+ */
+export function selectDailyViewData(tidelineData, startDate, endDate) {
```

Comment:
This function cognitive complexity is 17, a little refactor to simplify it would be better

Reply:
Sure it could be improved, but it will be better to simplify the data model which is far too complex for our needs.
I think changing that is out of scope of my PR.
I choose to only fix the problem I had with the "random" time range, and simplify the actual code, keeping the previous algorithm.
What I can do is to take out one inner function. But it won't really simplify it.